plugins/auth-node/src/sign-in/readDeclarativeSignInResolver.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Config } from '@backstage/config';
+import { JsonObject } from '@backstage/types';
+import { SignInResolver } from '../types';
+import { SignInResolverFactory } from './createSignInResolverFactory';
+
+/** @public */
+export interface ReadDeclarativeSignInResolverOptions<TAuthResult> {
+  config: Config;
+  signInResolverFactories: {
+    [name in string]: SignInResolverFactory<TAuthResult, unknown>;
+  };
+}
+
+/** @public */
+export function readDeclarativeSignInResolver<TAuthResult>(
+  options: ReadDeclarativeSignInResolverOptions<TAuthResult>,
+): SignInResolver<TAuthResult> | undefined {
+  const resolvers =
+    options.config
+      .getOptionalConfigArray('signIn.resolvers')
+      ?.map(resolverConfig => {
+        const resolverName = resolverConfig.getString('resolver');
+        if (!Object.hasOwn(options.signInResolverFactories, resolverName)) {
+          throw new Error(
+            `Sign-in resolver '${resolverName}' is not available`,
+          );
+        }
+        const resolver = options.signInResolverFactories[resolverName];
+        const { resolver: _ignored, ...resolverOptions } =
+          resolverConfig.get<JsonObject>();
+
+        return resolver(resolverOptions);
+      }) ?? [];
+
+  if (resolvers.length === 0) {
+    return undefined;
+  }
+
+  return async (profile, context) => {
+    for (const resolver of resolvers ?? []) {
+      try {
+        return await resolver(profile, context);
+      } catch (error) {
+        if (error?.name === 'NotFoundError') {
```

Comment:
Yeah I was a bit torn on this vs returning `undefined`. It would be more of a breaking change to handle `undefined` but might be more clear? On the other hand I figured this is relatively clear and happens to work for a lot of cases. I'll probably take a stab at moving over to `undefined` and see how bad it is.

Reply:
Took another look at this, and allowing sign-in resolvers to return `undefined` gets really awkward too. You end up with extra checks and loss of context of why the lookup failed. And there'd be a bunch of `catch (e) { if (e.name === 'NotFoundError') { return undefined } else throw e }` all over the place.

I really don't like this pattern but the alternative is also not great, leaning towards keeping it atm.

Reply:
Yeah agreed, I can definitely see that it's handy even though there's a small amount of "magic" to it, it doesn't really hurt